thread_management.cpp: Various Mandatory Threading Fixes | Resolve #398

lzardy opened the pull request:
- scePthreadMutexInit did not return default when the mutex attributes were empty, resulting in invalid accesses.
Now, we return default as expected.
- Current HeapAPI was incomplete.
Now it has a full struct with all 10 preset function replacement fields.
Additionally, we now account for the full size 10 array parameter passed to _sceKernelRtldSetApplicationHeapAPI.
- HeapAPI malloc replacement unaccounted for
Now, we at least have a fallback to `std::malloc`, such that allocation completes without early crash within user thread.

Resolves #398

lzardy committed:
Handle empty mutex attribute

- scePthreadMutexInit did not return default when the mutex attributes were empty, now it does

lzardy committed:
fix conditional unsafety

lzardy committed:
Update thread_management.cpp

fix deref

lzardy committed:
accurate heap api

- modified HeapAPI to a struct with preset function fields
- utilized the full array parameter passed to _sceKernelRtldSetApplicationHeapAPI

lzardy committed:
Merge branch 'main' of https://github.com/lzardy/shadPS4

lzardy committed:
Merge branch 'shadps4-emu:main' into main

lzardy committed:
Merge branch 'shadps4-emu:main' into main

lzardy committed:
fallback to std malloc

lzardy committed:
clang format

lzardy committed:
Declare all HeapAPI replacement functions

- calloc, realloc, memalign, reallocalign, malloc_stats, malloc_stats_fast, malloc_usable_size
- posix_memalign corrected parameters

lzardy committed:
Merge branch 'shadps4-emu:main' into main

lzardy committed:
Merge branch 'shadps4-emu:main' into main

raphaelthegreat commented on src/core/linker.h:
```diff
@@ -46,7 +46,21 @@ struct EntryParams {
+    PS4_SYSV_ABI size_t (*heap_malloc_usable_size)(void*);
+};
+
+typedef HeapAPI* AppHeapAPI;
```
```suggestion
using AppHeapAPI = HeapAPI*;
```

raphaelthegreat commented on src/core/linker.cpp:
```diff
@@ -335,10 +335,18 @@ void Linker::InitTlsForThread(bool is_primary) {
+            LOG_ERROR(Core_Linker, "TLS user malloc called, using std::malloc");
+            // TODO: ref rtld_tls_block_malloc, tls malloc replacement
+            // addr_out = heap_api->heap_malloc(total_tls_size);
             addr_out = std::malloc(total_tls_size);
```
Can this be limited to windows only? iirc linux has no such breakage and it's probably safer to use this. malloc on linux might return high addresses

lzardy committed:
resolve suggestions

- `using` definition replacement for AppHeapAPI
- linux uses heap_malloc, windows uses std::malloc
